### PR TITLE
use_relative_output_paths

### PR DESCRIPTION
**Pull request by mlin**
#604 #606

@rhpvorderman Here's a refactor based on my earlier comments, WDYT? (Opened a new PR because I didn't have permission to push to your branch, but keeping the commit history will ensure your co-authorship when we merge)

### CODE REVIEW

**Comment by rhpvorderman** [comment]
I did a test. Unfortunately, when using the call_cache it outputs everything in the same directory, without keeping the directory structure. This was not the case before the refactor. 
I am checking the code to see if this can be fixed easily.

File: WDL/runtime/task.py [hunk 1]
```diff
@@ -802,6 +818,61 @@ def map_paths(v: Value.Base, dn: str) -> Value.Base:
...
+                # target was generated by current task; use its path relative to the task work dir
+                if not os.path.basename(run_dir).startswith("download-"):  # except download tasks
+                    rel_link = os.path.relpath(real_target, os.path.join(run_dir, "work"))
+            elif path_really_within(target, run_dir):
+                # target is an out/ link generated by a call in the current workflow (task or
+                # subworkflow); use the link's path relative to that out/ dir
+                rel_link = os.path.relpath(target, run_dir)
+                p = rel_link.find("/out/")
+                rel_link = rel_link[p + 5 :] if p > 0 else None
+            # if neither of the above cases applies, then fall back to just the target basename
```
Comment (rhpvorderman) [inline WDL/runtime/task.py:852]: A case is missing for call cached entries. These also have an outdir. I wonder if it is possible to "know" that the code is taking a path from the cache.

Comment (mlin) [reply WDL/runtime/task.py:852]: @rhpvorderman Good catch, I'll look into this

**Comment by mlin** [comment]
@rhpvorderman updated, please give this a try? I think the logic is now very similar to your original, just with a few added heuristics to make it harder (but not impossible) to confuse it with other subdirectories named `work/` or `out/`.

**Comment by rhpvorderman** [comment]
@mlin, this works great! I tested it with call caching and output hardlinks, and both seem to work fine. Thanks :D!

**Comment by rhpvorderman** [comment]
@mlin, thanks again for your efforts!

**Pull request merged**